proposals/p2511.md

```diff
@@ -0,0 +1,190 @@
+classes by default. Leads issue
+[#686](https://github.com/carbon-language/carbon-lang/issues/686) gives some
+rules, but those rules are not part of this proposal.
+
+## Rationale
+
+-   [Language tools and ecosystem](/docs/project/goals.md#language-tools-and-ecosystem)
+    -   The values of variables can only change if either the address is taken,
+        including implicitly by `addr` self parameters, or at assignment
+        statements, making it easier to reason about where the value of a
+        variable can change in the control flow of a function.
+-   [Software and language evolution](/docs/project/goals.md#software-and-language-evolution)
+    -   This approach is conservative and can evolve to support assignment as a
+        subexpression.
+-   [Code that is easy to read, understand, and write](/docs/project/goals.md#code-that-is-easy-to-read-understand-and-write)
+    -   Assignments in subexpressions tend to be hard for humans to read and
+        understand. Disallowing them at the language level avoids the potential
+        for confusion, making code easier to read at the cost of making certain
+        constructs such as `n = arr[i++];` a little harder to write.
+    -   Easier to write a complete operator set because both `$` and `$=` can be
+        provided by implementing a single interface.
+-   [Practical safety and testing mechanisms](/docs/project/goals.md#practical-safety-and-testing-mechanisms)
+    -   Replaces correctness warning on `if (a = b)` with a language rule.
+-   [Interoperability with and migration from existing C++ code](/docs/project/goals.md#interoperability-with-and-migration-from-existing-c-code)
+    -   Providing largely the same set of symbols makes migration easier.
+    -   Some cost may be imposed by forcing a translation from assignment in
+        subexpressions to assignment as separate statements.
+
+## Alternatives considered
+
+### Allow assignment as a subexpression
+
+We could allow some or all forms of assignment as subexpressions, either with
+the same syntax or with some other syntax. For now, we are choosing to not
+pursue these options in order to gather data around how important they are. If
+we wish to allow assignment as a subexpression, we should consider adopting
+something like Python's "walrus operator" `variable := value` instead of using
+an `=` symbol.
+
+If we allow a three-term form of `for` statements, `for (init; cond; incr)`, we
+could allow assignment in the `incr` term. However, we currently do not support
+this syntax.
+
+### Allow chained assignment
+
+As a special case of assignment as a subexpression, we could allow chained
+assignment:
+
+```
+a = b = c = 0;
+```
+
+We could restrict this to only apply to simple assignment, not cases like
+`a += b -= c *= 2;`
+
+In leads issue [#451](https://github.com/carbon-language/carbon-lang/issues/451)
+it was decided that we would initially not support chained assignment.
```

Comment:
Should the rationale be provided here so that leads issues aren't a canonical reference for decisions?

Perhaps more importantly, what _was_ the rationale? What trade-offs did the leads consider? On a skim, [I see a decision](https://github.com/carbon-language/carbon-lang/issues/451#issuecomment-837483823), but without further context.

Reply:
Added some words, but honestly I'm not sure we had much rationale here other than that we didn't see sufficient motivation to add this feature compared to the cost.

Reply:
A lack of justification is still important to document. :)